[breaking] Remove auto detection of Arduino IDE built-in libraries and tools / Allow gRPC install of built-in libraries

cmaglie opened the pull request:
**Please check if the PR fulfills these requirements**

- [X] The PR has no duplicates (please search among the [Pull Requests](https://github.com/arduino/arduino-cli/pulls)
      before creating one)
- [X] The PR follows
      [our contributing guidelines](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/#pull-requests)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)
- [X] `UPGRADING.md` has been updated with a migration guide (for breaking changes)

* **What kind of change does this PR introduce?**
Remove the check for the CLI being bundled in the Arduino IDE

- **What is the current behavior?**
Arduino CLI checks if it's bundled with the Arduino IDE. This check allows the  CLI
to automatically use the libraries and tools bundled in the Arduino IDE

* **What is the new behavior?**
Now, this is not supported anymore unless the configuration keys `directories.builtin.libraries/tools` are set.

- **Does this PR introduce a breaking change, and is
[titled accordingly](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/#breaking)?**
Yes

* **Other information**:
Fix #1565

per1234 commented on configuration/directories.go:
```diff
@@ -24,12 +24,8 @@ import (
-		if bundledHardwareDir.IsDir() {
-			res.Add(bundledHardwareDir)
-		}
+	for _, bundledHardwareDir := range BuiltinToolsDirectories(Settings) {
+		res.Add(bundledHardwareDir)
```
We see here that this is more than a "tools" directory. It is a packages directory, which might contain platforms as well as tools.

I think this makes it a much more valuable feature, but it also makes the use of "tools" chosen here for the Go package API, configuration keys, and descriptions in the documentation inaccurate.

cmaglie replied:
Right, let's rename `*Tools*` -> `*Packages*`

cmaglie replied:
mmmm, actually no,  you can only have tools inside that path: https://github.com/arduino/arduino-cli/pull/1817/files#diff-bc91ebff69e54330f19e091571eca06927a59b29e43c53f6a4d5c04ee3dba243R41

```
	bundleToolDirs := configuration.BuiltinToolsDirectories(configuration.Settings)
	merr = append(merr, pm.LoadToolsFromBundleDirectories(bundleToolDirs)...)
```
so I think it's correct tools, and it's wrong the local variable name `bundledHardwareDir` that should be `bundledToolsDir`

per1234 replied:
```text
$ arduino-cli version
arduino-cli.exe  Version: git-snapshot Commit: eb6e52af Date: 2022-08-01T11:56:57Z

$ export ARDUINO_DIRECTORIES_DATA="/tmp/empty-directories-data"

$ arduino-cli core list

$ arduino-cli sketch new /tmp/FooSketch
Sketch created in: C:\Users\per\AppData\Local\Temp\FooSketch

$ arduino-cli compile -b arduino:avr:uno /tmp/FooSketch


Used platform Version Path


Error during build: Platform 'arduino:avr' not found: platform not installed
Try running `arduino-cli.exe core install arduino:avr`
```

As expected, no platforms were available because `directories.data` is empty.

Now I set `directories.builtin.tools` to point to a folder that contains packages and try again:

```text
$ export ARDUINO_DIRECTORIES_BUILTIN_TOOLS="C:/Users/per/AppData/Local/Arduino15/packages"

$ arduino-cli core list
ID          Installed Latest Name
arduino:avr 1.8.5     1.8.5  Arduino AVR Boards

$ arduino-cli compile -b arduino:avr:uno /tmp/FooSketch
Sketch uses 444 bytes (1%) of program storage space. Maximum is 32256 bytes.
Global variables use 9 bytes (0%) of dynamic memory, leaving 2039 bytes for local variables. Maximum is 2048 bytes.


Used platform Version Path
arduino:avr   1.8.5   C:\Users\per\AppData\Local\Arduino15\packages\arduino\hardware\avr\1.8.5
```

Voila! It used the platform from `directories.builtin.tools`

cmaglie replied:
Well... :-)

turns out that `configuration.HardwareDirectories(...)` should not return the bundled tools directory at all. So I'm removing these two lines. Nice catch @per1234!

cmaglie replied:
Another thing to consider is that the "bundled tools" supports a `builtin_tools_versions.txt` that describes the tools packager origin:

```go
	// We scan toolsPath content to find a "builtin_tools_versions.txt", if such file exists
	// then the all the tools are available in the same directory, mixed together, and their
	// name and version are written in the "builtin_tools_versions.txt" file.
	// If no "builtin_tools_versions.txt" is found, then the directory structure is the classic
	// TOOLSPATH/TOOL-NAME/TOOL-VERSION and it will be parsed as such and associated to an
	// "unnamed" packager.

	// TODO: get rid of "builtin_tools_versions.txt"
```

this is a very Arduino IDE 1.8.x legacy, where we have this `builtin_tools_versions.txt` inside an **hardware** folder:

```
$ tree ~/Downloads/arduino-1.8.13/hardware
hardware
├── arduino
│   └── avr
│       ├── boards.txt
│       ├── cores
│       ├── libraries
│       │   ├── EEPROM
│       │   ├── HID
│       │   ├── SoftwareSerial
│       │   ├── SPI
│       │   └── Wire
│       ├── platform.txt
│       ├── programmers.txt
│       └── variants
├── package_index_bundled.json
└── tools
    └── avr
        ├── avr
        │   ├── bin
        │   ├── include
        │   └── lib
        ├── bin
        │   ├── arduinoOTA
        │   ├── avr-c++
        │   ├── avr-c++filt
        │   ├── avr-cpp
        │   ├── avrdude
        │   ├── avr-g++
        │   .....
        ├── builtin_tools_versions.txt
        ,,,
```

I'm not sure that we can remove this "feature" now, but removing the "bundled tools dir" from the "hardware dirs list" should not produce any visible effect on the `arduino-builder` used in the Arduino IDE 1.8.x.

cmaglie replied:
https://github.com/arduino/arduino-cli/pull/1817/commits/696ca8212397c0981f11a55c7f02eccc73609bbd

per1234 replied:
>```go
> the directory structure is the classic
>	// TOOLSPATH/TOOL-NAME/TOOL-VERSION and it will be parsed as such and associated to an
>	// "unnamed" packager.
>```

I guess this is the explanation why the wrong `teensy:teensy-compile@1.56.1` tool is used instead of the expected `arduino:avr-gcc@7.3.0-atmel3.6.1-arduino7` while compiling for `arduino:avr:uno` when I set `directories.builtin.tools` to the `packages` subfolder of my normal `directories.data` folder, and shouldn't be cause for concern?:

```text
$ export ARDUINO_DIRECTORIES_DATA="/tmp/empty-directories-data2"

$ export ARDUINO_DIRECTORIES_BUILTIN_TOOLS="C:/Users/per/AppData/Local/Arduino15/packages"

$ ./arduino-cli core install arduino:avr
Tool arduino:avr-gcc@7.3.0-atmel3.6.1-arduino7 already installed
Tool arduino:avrdude@6.3.0-arduino17 already installed
Tool arduino:arduinoOTA@1.3.0 already installed
Downloading packages...
arduino:avr@1.8.5 already downloaded
Installing platform arduino:avr@1.8.5...
Configuring platform....
Platform arduino:avr@1.8.5 installed

$ ./arduino-cli compile -b arduino:avr:uno -v PERTILLISCH-FooSketch/

Using board 'uno' from platform in folder: C:\Users\per\AppData\Local\Temp\empty-directories-data2\packages\arduino\hardware\avr\1.8.5
Using core 'arduino' from platform in folder: C:\Users\per\AppData\Local\Temp\empty-directories-data2\packages\arduino\hardware\avr\1.8.5
Detecting libraries used...
"C:\\Users\\per\\AppData\\Local\\Arduino15\\packages\\teensy\\tools\\teensy-compile\\1.56.1\\avr/bin/avr-g++" -c -g -Os -w -std=gnu++11 -fpermissive -fno-exceptions -ffunction-sections -fdata-sections -fno-threadsafe-statics -Wno-error=narrowing -flto -w -x c++ -E -CC -mmcu=atmega328p -DF_CPU=16000000L -DARDUINO=10607 -DARDUINO_AVR_UNO -DARDUINO_ARCH_AVR "-IC:\\Users\\per\\AppData\\Local\\Temp\\empty-directories-data2\\packages\\arduino\\hardware\\avr\\1.8.5\\cores\\arduino" "-IC:\\Users\\per\\AppData\\Local\\Temp\\empty-directories-data2\\packages\\arduino\\hardware\\avr\\1.8.5\\variants\\standard" "C:\\Users\\per\\AppData\\Local\\Temp\\arduino-sketch-A91912F4E8442FE7AB972C142696BA39\\sketch\\PERTILLISCH-FooSketch.ino.cpp" -o nul
Error while detecting libraries included by C:\Users\per\AppData\Local\Temp\arduino-sketch-A91912F4E8442FE7AB972C142696BA39\sketch\PERTILLISCH-FooSketch.ino.cpp
Generating function prototypes...
"C:\\Users\\per\\AppData\\Local\\Arduino15\\packages\\teensy\\tools\\teensy-compile\\1.56.1\\avr/bin/avr-g++" -c -g -Os -w -std=gnu++11 -fpermissive -fno-exceptions -ffunction-sections -fdata-sections -fno-threadsafe-statics -Wno-error=narrowing -flto -w -x c++ -E -CC -mmcu=atmega328p -DF_CPU=16000000L -DARDUINO=10607 -DARDUINO_AVR_UNO -DARDUINO_ARCH_AVR "-IC:\\Users\\per\\AppData\\Local\\Temp\\empty-directories-data2\\packages\\arduino\\hardware\\avr\\1.8.5\\cores\\arduino" "-IC:\\Users\\per\\AppData\\Local\\Temp\\empty-directories-data2\\packages\\arduino\\hardware\\avr\\1.8.5\\variants\\standard" "C:\\Users\\per\\AppData\\Local\\Temp\\arduino-sketch-A91912F4E8442FE7AB972C142696BA39\\sketch\\PERTILLISCH-FooSketch.ino.cpp" -o "C:\\Users\\per\\AppData\\Local\\Temp\\arduino-sketch-A91912F4E8442FE7AB972C142696BA39\\preproc\\ctags_target_for_gcc_minus_e.cpp"
avr-g++: error: device-specs/specs-atmega328p: No such file or directory
```

cmaglie commented:
More work is going into this PR: the handling of the historic Arduino IDE bundled libraries will be implemented in the CLI.

cmaglie commented:
The gRPC `LibraryInstall` now has an extra field called `install_location` that can have two values:
* `LIBRARY_INSTALL_LOCATION_USER` (this is the default if not set)
* `LIBRARY_INSTALL_LOCATION_BUILTIN`

This change allows the gRPC client (IDE) to install libraries in the "builtin libraries" directory.
The "builtin libraries" directory can be set via the CLI configuration key `directories.builtin.libraries`.

per1234 commented on configuration/defaults.go:
```diff
@@ -65,5 +65,7 @@ func SetDefaults(settings *viper.Viper) {
 	settings.BindEnv("directories.Downloads", "ARDUINO_DOWNLOADS_DIR")
 	settings.BindEnv("directories.Data", "ARDUINO_DATA_DIR")
+	settings.BindEnv("directories.builtin.tools", "ARDUINO_BUILTIN_TOOLS_DIR")
+	settings.BindEnv("directories.builtin.libraires", "ARDUINO_BUILTIN_LIBRARIES_DIR")
```
```suggestion
	settings.BindEnv("directories.builtin.libraries", "ARDUINO_BUILTIN_LIBRARIES_DIR")
```
Fix typo in key name.

But more importantly, are these environment variables even needed? From the comment in this file, it seems the purpose is for backwards compatibility with some previous version that apparently used them. But that doesn't seem to be the case with these two environment variables from what I can tell.

If they are not needed for backwards compatibility, I recommend removing them. The documented and standard auto-generated environment variables (e.g., `ARDUINO_DIRECTORIES_BUILTIN_LIBRARIES`) can be used instead.

cmaglie replied:
Right, better remove them.

per1234 commented on docs/configuration.md:
```diff
@@ -9,6 +9,11 @@
+    user to install them, but with the lowest priority over other installed libraries with the same name, it's the
+    equivalent of the Arduino IDE's bundled libraries directory.
+  - `builtin.tools` - it's a list of directories of tools that will be available to all platforms without the need for
+    the user to install them, it's the equivalent of the Arduino IDE's bundled tools directory.
```
```suggestion
    the user to install them, it's the equivalent of the Arduino IDE 1.x bundled tools directory.
```
I think it is worth clarifying this since some readers will likely not be familiar with this "bundled tools directory". They might otherwise be confused after trying to learn about it by studying Arduino IDE 2.x, which has no such thing.

kittaakos commented:
As reported in IDE2 code [here](https://github.com/arduino/arduino-ide/pull/1370#pullrequestreview-1091316293), it does not work.

```
root ERROR Request install failed with error: 2 UNKNOWN: Builtin libraries directory not set Error: 2 UNKNOWN: Builtin libraries directory not set
    at Object.callErrorFromStatus (/Users/a.kitta/dev/git/arduino-ide/node_modules/@grpc/grpc-js/build/src/call.js:31:26)
    at Object.onReceiveStatus (/Users/a.kitta/dev/git/arduino-ide/node_modules/@grpc/grpc-js/build/src/client.js:349:49)
    at Object.onReceiveStatus (/Users/a.kitta/dev/git/arduino-ide/node_modules/@grpc/grpc-js/build/src/client-interceptors.js:328:181)
    at /Users/a.kitta/dev/git/arduino-ide/node_modules/@grpc/grpc-js/build/src/call-stream.js:187:78
    at processTicksAndRejections (node:internal/process/task_queues:78:11)
```

Why should the IDE2 set any path? We discussed that the default `install_location` which is `LIBRARY_INSTALL_LOCATION_USER` will install to `#directories/user/libraries`, the `LIBRARY_INSTALL_LOCATION_BUILTIN` will install into `#directories/data/libraries.

cmaglie commented:
Uhm... shall we make it default to `.../Arduino15/libraries`?

The rationale for this PR is that I see the "bundled" libraries as a legacy that we inherited from Arduino IDE 1.8.x, and I would like to separate them from the pure-CLI users. The bundled libraries can't be managed from the command line (there is no `lib install --bundled` flags right now). If we make the "bundled dir" available by default, the command-line user will see these "bundled" libraries appearing after the first run of the Arduino IDE.

@ubidefeo @per1234 any comments on this?

> Why should the IDE2 set any path? 

The IDE should be already setting the sketchbook path AFAIK, the bundled libs would be another path to set. But I'm open to defaulting it to `$DATA_DIR/libraries` if we decide to do so.

kittaakos commented:
> > Why should the IDE2 set any path?
> 
> The IDE should be already setting the sketchbook path AFAIK

No, IDE2 is not setting anything. There is the default CLI config that is generated by the CLI when IDE2 starts, and it's missing. The sketchbook path is inferred from the CLI config: `#directories/user`. Whenever a user wants to modify the sketchbook path from IDE2, eventually, the CLI config's `#directories/user` location will change and IDE2 restarts the daemon.

cmaglie commented:
Another solution may be to set the default path only when the CLI is running as a daemon, this way:
* command-line users will not have to handle the weird "bundled" libraries
* the IDE would not need to set up paths

cmaglie commented:
I've implemented the default for the bundled library directory when running as a daemon, as I said in my last comment.

per1234 reviewed (changes requested):
### Describe the problem

[`cc.arduino.cli.commands.v1.ArduinoCoreService.LibraryUninstall`](https://arduino.github.io/arduino-cli/dev/rpc/commands/#libraryuninstallrequest) uninstalls the library that was installed to `directories.builtin.libraries`
### To reproduce

#### Set up

```text
$ arduino-cli version
arduino-cli.exe  Version: git-snapshot Commit: fb57a522 Date: 2022-08-31T14:09:52Z

$ arduino-cli lib install Servo
Downloading Servo@1.1.8...
Servo@1.1.8 already downloaded
Installing Servo@1.1.8...
Installed Servo@1.1.8

$ arduino-cli daemon --debug
Daemon is now listening on 127.0.0.1:50051
```

#### Demo

Use [**grpcurl**](https://github.com/fullstorydev/grpcurl) to run the following commands in another terminal:

```text
$ grpcurl \
  -plaintext \
  -import-path ./rpc \
  -proto cc/arduino/cli/commands/v1/commands.proto \
  127.0.0.1:50051 \
  cc.arduino.cli.commands.v1.ArduinoCoreService.Create

{
  "instance": {
    "id": 1
  }
}

$ grpcurl \
  -plaintext \
  -import-path ./rpc \
  -proto cc/arduino/cli/commands/v1/commands.proto \
  -d '{"instance": {"id": 1}}' \
  127.0.0.1:50051 \
  cc.arduino.cli.commands.v1.ArduinoCoreService.Init

$ grpcurl \
  -plaintext \
  -import-path ./rpc \
  -proto cc/arduino/cli/commands/v1/commands.proto \
  -d '{"instance": {"id": 1}, "name": "Servo", "install_location": 1}' \
  127.0.0.1:50051 \
  cc.arduino.cli.commands.v1.ArduinoCoreService.LibraryInstall

{
  "taskProgress": {
    "name": "Downloading Servo@1.1.8"
  }
}
{
  "progress": {
    "file": "Servo@1.1.8",
    "completed": true
  }
}
{
  "taskProgress": {
    "completed": true
  }
}
{
  "taskProgress": {
    "name": "Installing Servo@1.1.8"
  }
}
{
  "taskProgress": {
    "message": "Installed Servo@1.1.8",
    "completed": true
  }
}
{

}

$ grpcurl \
  -plaintext \
  -import-path ./rpc \
  -proto cc/arduino/cli/commands/v1/commands.proto \
  -d '{"instance": {"id": 1}, "all": true, "name": "Servo"}' \
  127.0.0.1:50051 \
  cc.arduino.cli.commands.v1.ArduinoCoreService.LibraryList

{
  "installedLibraries": [
    {
      "library": {
        "name": "Servo",
        "author": "Michael Margolis, Arduino",
        "maintainer": "Arduino \u003cinfo@arduino.cc\u003e",
        "sentence": "Allows Arduino boards to control a variety of servo motors.",
        "paragraph": "This library can control a great number of servos.\u003cbr /\u003eIt makes careful use of timers: the library can control 12 servos using only 1 timer.\u003cbr /\u003eOn the Arduino Due you can control up to 60 servos.",
        "website": "https://www.arduino.cc/reference/en/libraries/servo/",
        "category": "Device Control",
        "architectures": [
          "avr",
          "megaavr",
          "sam",
          "samd",
          "nrf52",
          "stm32f4",
          "mbed",
          "mbed_nano",
          "mbed_portenta",
          "mbed_rp2040"
        ],
        "installDir": "C:\\Users\\per\\AppData\\Local\\Arduino15\\libraries\\Servo",
        "sourceDir": "C:\\Users\\per\\AppData\\Local\\Arduino15\\libraries\\Servo\\src",
        "realName": "Servo",
        "version": "1.1.8",
        "license": "Unspecified",
        "layout": "LIBRARY_LAYOUT_RECURSIVE",
        "examples": [
          "C:\\Users\\per\\AppData\\Local\\Arduino15\\libraries\\Servo\\examples\\Knob",
          "C:\\Users\\per\\AppData\\Local\\Arduino15\\libraries\\Servo\\examples\\Sweep"
        ],
        "providesIncludes": [
          "Servo.h"
        ]
      }
    },
    {
      "library": {
        "name": "Servo",
        "author": "Michael Margolis, Arduino",
        "maintainer": "Arduino \u003cinfo@arduino.cc\u003e",
        "sentence": "Allows Arduino boards to control a variety of servo motors.",
        "paragraph": "This library can control a great number of servos.\u003cbr /\u003eIt makes careful use of timers: the library can control 12 servos using only 1 timer.\u003cbr /\u003eOn the Arduino Due you can control up to 60 servos.",
        "website": "https://www.arduino.cc/reference/en/libraries/servo/",
        "category": "Device Control",
        "architectures": [
          "avr",
          "megaavr",
          "sam",
          "samd",
          "nrf52",
          "stm32f4",
          "mbed",
          "mbed_nano",
          "mbed_portenta",
          "mbed_rp2040"
        ],
        "installDir": "C:\\Users\\per\\Documents\\Arduino\\libraries\\Servo",
        "sourceDir": "C:\\Users\\per\\Documents\\Arduino\\libraries\\Servo\\src",
        "realName": "Servo",
        "version": "1.1.8",
        "license": "Unspecified",
        "location": "LIBRARY_LOCATION_USER",
        "layout": "LIBRARY_LAYOUT_RECURSIVE",
        "examples": [
          "C:\\Users\\per\\Documents\\Arduino\\libraries\\Servo\\examples\\Knob",
          "C:\\Users\\per\\Documents\\Arduino\\libraries\\Servo\\examples\\Sweep"
        ],
        "providesIncludes": [
          "Servo.h"
        ]
      }
    }
  ]
}

$ grpcurl \
  -plaintext \
  -import-path ./rpc \
  -proto cc/arduino/cli/commands/v1/commands.proto \
  -d '{"instance": {"id": 1}, "name": "Servo", "version": "1.1.8"}' \
  127.0.0.1:50051 \
  cc.arduino.cli.commands.v1.ArduinoCoreService.LibraryUninstall

{
  "taskProgress": {
    "name": "Uninstalling Servo@1.1.8"
  }
}
{
  "taskProgress": {
    "completed": true
  }
}
{

}

$ grpcurl \
  -plaintext \
  -import-path ./rpc \
  -proto cc/arduino/cli/commands/v1/commands.proto \
  -d '{"instance": {"id": 1}, "all": true, "name": "Servo"}' \
  127.0.0.1:50051 \
  cc.arduino.cli.commands.v1.ArduinoCoreService.LibraryList

{
  "installedLibraries": [
    {
      "library": {
        "name": "Servo",
        "author": "Michael Margolis, Arduino",
        "maintainer": "Arduino \u003cinfo@arduino.cc\u003e",
        "sentence": "Allows Arduino boards to control a variety of servo motors.",
        "paragraph": "This library can control a great number of servos.\u003cbr /\u003eIt makes careful use of timers: the library can control 12 servos using only 1 timer.\u003cbr /\u003eOn the Arduino Due you can control up to 60 servos.",
        "website": "https://www.arduino.cc/reference/en/libraries/servo/",
        "category": "Device Control",
        "architectures": [
          "avr",
          "megaavr",
          "sam",
          "samd",
          "nrf52",
          "stm32f4",
          "mbed",
          "mbed_nano",
          "mbed_portenta",
          "mbed_rp2040"
        ],
        "installDir": "C:\\Users\\per\\Documents\\Arduino\\libraries\\Servo",
        "sourceDir": "C:\\Users\\per\\Documents\\Arduino\\libraries\\Servo\\src",
        "realName": "Servo",
        "version": "1.1.8",
        "license": "Unspecified",
        "location": "LIBRARY_LOCATION_USER",
        "layout": "LIBRARY_LAYOUT_RECURSIVE",
        "examples": [
          "C:\\Users\\per\\Documents\\Arduino\\libraries\\Servo\\examples\\Knob",
          "C:\\Users\\per\\Documents\\Arduino\\libraries\\Servo\\examples\\Sweep"
        ],
        "providesIncludes": [
          "Servo.h"
        ]
      }
    }
  ]
}
```

🐛 The library in `LIBRARY_LOCATION_BUILTIN` was uninstalled instead of the library in `LIBRARY_LOCATION_USER`

### Expected behavior

`cc.arduino.cli.commands.v1.ArduinoCoreService.LibraryUninstall` only uninstalls libraries from `directories.user`

### Arduino CLI version

fb57a522

### Operating system

Windows 10

### Additional context

The bug does not occur if `"version": "1.1.8"` is removed from the `cc.arduino.cli.commands.v1.ArduinoCoreService.LibraryUninstall` request.

cmaglie commented:
> :bug: The library in LIBRARY_LOCATION_BUILTIN was uninstalled instead of the library in LIBRARY_LOCATION_USER

@per1234 I don't know how you manage to find these edge cases but it's amazing!

In this case, `LibraryUninstall` shows the problem only when it's called with the library version parameter. I've fixed and added the test.

per1234 reviewed (approved):
I verified that the bug reported in [my previous review](https://github.com/arduino/arduino-cli/pull/1817#pullrequestreview-1092021620) was fixed by https://github.com/arduino/arduino-cli/pull/1817/commits/ad98097a51dda1a196d8c25e483e0710ff58172b

Excellent work on this Cristian!
